tests/testsuite/member_discovery.rs

```diff
@@ -0,0 +1,36 @@
+//! Tests for workspace member discovery.
+
+//use cargo::core::resolver::ResolveError;
+use cargo::core::{Shell, Workspace};
+use cargo::ops;
+use cargo::util::config::Config;
+
+use cargo_test_support::install::cargo_home;
+use cargo_test_support::project;
+use cargo_test_support::registry;
+
+/// Tests exclusion of non-directory files from workspace member discovery using glob `*`.
+#[cargo_test]
+fn bad_file_member_exclusion() {
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+            [workspace]
+            members = [ "crates/*" ]
+        "#,
+        )
+        .file("crates/.DS_Store", "PLACEHOLDER")
+        .build();
+
+    // Prevent this test from accessing the network by setting up .cargo/config.
+    registry::init();
+    let config = Config::new(
+        Shell::from_write(Box::new(Vec::new())),
+        cargo_home(),
+        cargo_home(),
+    );
+    let ws = Workspace::new(&p.root().join("Cargo.toml"), &config).unwrap();
+    let (pkg_set, _resolve) = ops::resolve_ws(&ws).unwrap();
+    assert_eq!(pkg_set.package_ids().count(), 0);
```

Comment:
I'm a bit uncertain why this is using the resolver to validate the members.  Could this possibly just call `ws.members()` to check the set of members found?  That would also remove the need for the non-default config and the `registry::init`.

Reply:
Thanks for pointing. it to me. 
It's just because I am not familiar with Cargo codebase. I'll fix it.